---
title: "Distance of Msi2 HyperTRIBE edited site in MOLM13 from the nearest Msi2 iCLIP motif in NB4"
output: html_document
---

Purpose: We want to see how close Msi2 HyperTRIBE edited site in MOLM13 compares to Msi2 iCLIP motif in NB4.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r Zining's code: library load}
require(GenomicFeatures)
require(ggplot2)
require(openxlsx)
require(org.Hs.eg.db)
require(TxDb.Hsapiens.UCSC.hg19.knownGene)
require(VennDiagram)
```


```{r Zining's code: load input data}
setwd("/Users/chuk/Documents/MSI2-hyperTRIBE/MOLM13/")

# load transcript database
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene

# get entrez id -> gene name mapping
x <- org.Hs.eg.db::org.Hs.egSYMBOL
mapped_genes <- mappedkeys(x)
xx <- as.list(x[mapped_genes])

# load genes (returns GRanges object)
hg19.genes <- GenomicFeatures::genes(txdb)

# start from GenomicRanges (code for making this on lilac cluster)
peak.data.fname <- file.path("/Users/chuk/Documents/MSI2-hyperTRIBE/MOLM13/data/peakdata.rds")
gr.fname <- file.path("/Users/chuk/Documents/MSI2-hyperTRIBE/MOLM13/data/CLIP_genomic_ranges.RData")

# molm13 raw dataframe (for unfiltered data)
molm13.edit.fname <- file.path("/Users/chuk/Documents/MSI2-hyperTRIBE/MOLM13/data/molm13_snp_counts_dedupped_significant.csv")
```

Below is Zi-Ning's code. Parts were taken out because the input files were not provided. Instead, Diu suggested I load another set of input files that seems to be downstream of this code and reproduces Zi-Ning's plots accurately.

```{r Zining's code: load data and filter MOLM13 based on fpkm and diff.freq}
# functions
CalculateDistance <- function(query.idx, subject.idx, strand) {
  if(length(query.idx) == 0 || length(subject.idx) == 0 || length(strand) == 0) {
    return(NA)
  }
  if(as.character(strand)=="+"){
    return(query.idx - subject.idx)
  } else {
    return(subject.idx - query.idx)
  }
}

# loads the following Granges:
# k562.signif.gr
# molm13.signif.gr
# nb4.signif.gr
# peaks.signif.gr
load(gr.fname)

# peaks.data contains peaks found by CLIPanalyze
peaks.data <- readRDS(peak.data.fname)

# dataframe of molm13 HyperTRIBE hits
molm13.edit.df <- read.csv(molm13.edit.fname)

# apply new filters
ada.fpkm.thres <- 5
dcd.fpkm.thres <- 5
mig.fpkm.thres <- 5
diff.freq.thres <- 0.1

# subset molm13 dataframe based on new filters (previously only applied ada.fpkm and diff.freq filters)
molm13.filtered.df <- subset(molm13.edit.df, 
                             ADA.fpkm > ada.fpkm.thres &
                               DCD.fpkm > dcd.fpkm.thres &
                               MIG.fpkm > mig.fpkm.thres &
                               diff.frequency > diff.freq.thres)

# create Genomic Ranges object of MOLM13 data after more stringent filters
molm13.filtered.df$start <- molm13.filtered.df$pos
molm13.filtered.df$end <- molm13.filtered.df$pos
molm13.filtered.gr <- GenomicRanges::makeGRangesFromDataFrame(molm13.filtered.df)

# peaks small
peaks.small.gr <- peaks.data$peaks[order(peaks.data$peaks@elementMetadata$padj)]
peaks.small.gr <- peaks.small.gr[1:3000]

#molm13 small
molm13.small.df <- molm13.filtered.df[order(molm13.filtered.df$p.adj),]
molm13.small.df <- molm13.small.df[!duplicated(molm13.small.df$entrez.id),]
```

Diu suggested me to upload the 3 csv files.
Below is Zi-Ning's code to plot venn diagrams, which (to my best guess) shows the Msi2 gene targets shared between NB4 and MOLM13, and K562 and MOLM13. Then Zi-Ning identifies nearest neighboring CLIP peaks and calculates their distance.

```{r Zi-Ning's code: venn diagrams}
nb4.hits.df <- read.csv("/Users/chuk/Documents/MSI2-hyperTRIBE/MOLM13/data/nb4_short_genes.csv")
k562.hits.df <- read.csv("/Users/chuk/Documents/MSI2-hyperTRIBE/MOLM13/data/k562_short_genes.csv")
molm13.small.df <- read.csv("/Users/chuk/Documents/MSI2-hyperTRIBE/MOLM13/data/molm13_short_genes.csv")

# NOTE CHANGED MOLM13 GENE IDS - Karen: changed filename and changed nb4.hits.gr to nb4.hits.df
nb4.molm13.vd <- VennDiagram::venn.diagram(
  x = list(
    "NB4" = nb4.hits.df$gene_id,
    "MOLM13" = molm13.small.df$entrez.id
  ),
  filename = file.path("nb4_molm13_venn.tiff"),
  col = "transparent",
  fill = c("cornflowerblue", "green"),
  alpha = 0.50,
  label.col = rep("blue", 3),
  scaled = TRUE,
  ext.text = TRUE,
  ext.line.lwd = 2,
  ext.dist = -0.15,
  ext.length = 0.9,
  ext.pos = -4,
  inverted = FALSE,
  cex = 1.5,
  fontfamily = "sans",
  fontface = "bold",
  cat.col = c("darkblue", "darkgreen"),
  cat.cex = 1.5,
  cat.fontfamily = "sans",
  rotation.degree = 0
)

k562.molm13.vd <- VennDiagram::venn.diagram(
  x = list(
    "K562" = k562.hits.df$gene_id,
    "MOLM13" = molm13.small.df$entrez.id
  ),
  filename = file.path("k562_molm13_venn.tiff"),
  col = "transparent",
  fill = c("cornflowerblue", "green"),
  alpha = 0.50,
  label.col = rep("blue", 3),
  scaled = TRUE,
  ext.text = TRUE,
  ext.line.lwd = 2,
  ext.dist = -0.15,
  ext.length = 0.9,
  ext.pos = -4,
  inverted = FALSE,
  cex = 1.5,
  fontfamily = "sans",
  fontface = "bold",
  cat.col = c("darkblue", "darkgreen"),
  cat.cex = 1.5,
  cat.fontfamily = "sans",
  rotation.degree = 0
)

# get indices of nearest neighbors
molm13.nb4.nearest <- GenomicRanges::nearest(molm13.signif.gr, nb4.signif.gr)
molm13.k562.nearest <- GenomicRanges::nearest(molm13.signif.gr, k562.signif.gr)
molm13.peaks.nearest <- GenomicRanges::nearest(molm13.signif.gr, peaks.signif.gr)
molm13.peaks.unfiltered.nearest <- GenomicRanges::nearest(molm13.signif.gr, peaks.data$peaks)

# get indices of nearest neighbors (the same shit for more stringently filtered data)
molm13.filtered.nb4.nearest <- GenomicRanges::nearest(molm13.filtered.gr, nb4.signif.gr)
molm13.filtered.k562.nearest <- GenomicRanges::nearest(molm13.filtered.gr, k562.signif.gr)
molm13.filtered.peaks.nearest <- GenomicRanges::nearest(molm13.filtered.gr, peaks.data$peaks)

# replace all NA values with 0
molm13.nb4.nearest[is.na(molm13.nb4.nearest)] <- 0
molm13.k562.nearest[is.na(molm13.k562.nearest)] <- 0
molm13.peaks.nearest[is.na(molm13.peaks.nearest)] <- 0

# reaplce all NA values with 0 (same shit for more stringently filtered data)
molm13.filtered.nb4.nearest[is.na(molm13.filtered.nb4.nearest)] <- 0
molm13.filtered.k562.nearest[is.na(molm13.filtered.k562.nearest)] <- 0
molm13.filtered.peaks.nearest[is.na(molm13.filtered.peaks.nearest)] <- 0

# get NB4/K562 distances (publicly available datasets)
molm13.nb4.distance <- mapply(function(q.idx, s.idx, strand) CalculateDistance(q.idx, s.idx, strand),
                              as.list(molm13.signif.gr@ranges@start),
                              as.list(nb4.signif.gr[c(molm13.nb4.nearest)]@ranges@start),
                              as.list(molm13.signif.gr@strand))
# Commented out because using new CLIPanalyze pipeline
# molm13.k562.distance <- mapply(function(q.idx, s.idx, strand) CalculateDistance(q.idx, s.idx, strand),
#                               as.list(molm13.signif.gr@ranges@start),
#                               as.list(k562.signif.gr[c(molm13.k562.nearest)]@ranges@start),
#                               as.list(molm13.signif.gr@strand))

# K562 peaks distance (from CLIPanalyze)
molm13.peaks.distance <- mapply(function(q.idx, s.idx, strand) CalculateDistance(q.idx, s.idx, strand),
                                as.list(molm13.signif.gr@ranges@start),
                                as.list(peaks.data$peaks[c(molm13.peaks.unfiltered.nearest)]@ranges@start),
                                as.list(molm13.signif.gr@strand))

# again for more stringent filters (LOL this code is horrible.)
# get NB4/K562 distances (publicly available datasets)
molm13.filtered.nb4.distance <- mapply(function(q.idx, s.idx, strand) CalculateDistance(q.idx, s.idx, strand),
                              as.list(molm13.filtered.gr@ranges@start),
                              as.list(nb4.signif.gr[c(molm13.filtered.nb4.nearest)]@ranges@start),
                              as.list(molm13.filtered.gr@strand))
# Commented out because using new CLIPanalyze pipeline
# molm13.filtered.k562.distance <- mapply(function(q.idx, s.idx, strand) CalculateDistance(q.idx, s.idx, strand),
#                                as.list(molm13.filtered.gr@ranges@start),
#                                as.list(k562.signif.gr[c(molm13.filtered.k562.nearest)]@ranges@start),
#                                as.list(molm13.filtered.gr@strand))

# K562 peaks distance (from CLIPanalyze)
molm13.filtered.peaks.distance <- mapply(function(q.idx, s.idx, strand) CalculateDistance(q.idx, s.idx, strand),
                                as.list(molm13.filtered.gr@ranges@start),
                                as.list(peaks.data$peaks[c(molm13.filtered.peaks.nearest)]@ranges@start),
                                as.list(molm13.filtered.gr@strand))

# analysis of K562 peaks vs NB4 peaks (K562 peaks called with CLIPanalyze)
k562.nb4.dist <- GenomicRanges::distanceToNearest(peaks.data$peaks, nb4.signif.gr)
k562.nb4.df <- data.frame(chr = peaks.data$peaks@seqnames,
                          pos = peaks.data$peaks@ranges@start,
                          dist = k562.nb4.dist@elementMetadata$distance)
k562.nb4.hist <- ggplot(data=k562.nb4.df, aes(k562.nb4.df$dist)) + 
  geom_histogram(breaks=seq(0,4000, by=20),
                 col="white", 
                 fill="blue", 
                 alpha=0.2) + 
  labs(title="Distance from K562 peak to nearest NB4 CLIP site") +
  labs(x="Distance (bp)") +
  labs(y="Count")


# create df with all of the distances
distance.df <- data.frame(chr <- molm13.signif.gr@seqnames,
                          strand <- molm13.signif.gr@strand,
                          pos <- molm13.signif.gr@ranges@start,
                          nb4.dist <- molm13.nb4.distance,
                          # k562.dist <- molm13.k562.distance,
                          peaks.dist <- molm13.peaks.distance)

# create df with all of the distances (filtered MOLM13)
filtered.distance.df <- data.frame(chr <- molm13.filtered.gr@seqnames,
                          strand <- molm13.filtered.gr@strand,
                          pos <- molm13.filtered.gr@ranges@start,
                          nb4.dist <- molm13.filtered.nb4.distance,
                          # k562.dist <- molm13.filtered.k562.distance,
                          peaks.dist <- molm13.filtered.peaks.distance)
```



```{r Zi-Ning's code: Histogram, Karen's code: PDF plot and probability calculation}
# NB4histogram
nb4.hist <- ggplot(data=distance.df, aes(distance.df$nb4.dist)) + 
  geom_histogram(breaks=seq(-1000,1000, by=50), 
                 col="white", 
                 fill="blue", 
                 alpha=0.2) + 
  labs(title="Distance to nearest NB4 CLIP site") +
  labs(x="Distance (bp)") +
  labs(y="Count")

# Karen's code below:
nb4.hist # Karen: to plot histogram

pdf("/Users/chuk/Documents/MSI2-hyperTRIBE/MOLM13/figures/Distance_to_nearest_NB4_CLIP_site.pdf", 15, 10, useDingbats = F )
ggplot(distance.df, aes(x=distance.df$nb4.dist)) + geom_density(fill="dodgerblue") +
  scale_x_continuous(limits = c(-1000, 1000)) +
  labs(title="Distance to nearest NB4 CLIP site\n") +
  labs(x="\nDistance (bp)") +
  labs(y="Density\n") +
  theme_classic() +
  theme(plot.title = element_text(size=40), 
        axis.text=element_text(size=40), 
        axis.text.x = element_text(angle = 90),
        axis.title.x = element_text(size=40), 
        axis.title.y = element_text(size=40))
dev.off()

saveRDS(distance.df, "/Users/chuk/Documents/MSI2-hyperTRIBE/MOLM13/data/distance_to_nearest_NB4_clip_site-KarenChu.rds")

# Calculate probability of finding Msi2 HyperTRIBE edited site 500 bp away from nearest NB4 CLIP site
ecdf.nb4 <- ecdf(distance.df$nb4.dist)
ecdf.nb4.values <- data.frame(x = environment(ecdf.nb4)$x, y = environment(ecdf.nb4)$y)
ecdf.interval.250 <- subset(ecdf.nb4.values, x <=500) # +500 to negative infinity
ecdf.interval.neg250 <- subset(ecdf.nb4.values, x < -500) # -500 to negative infinity

probability.pos250.to.neg250 <- (ecdf.interval.250[ nrow(ecdf.interval.250),]$y) - (ecdf.interval.neg250[ nrow(ecdf.interval.neg250),]$y) #Calculates probability from -250 <= X <= +250.

cat("Probability of finding a Msi2 HyperTribe edited site 500 bp away from NB4 CLIP site in MOLM13 is ", probability.pos250.to.neg250)
```

















